### PR TITLE
Don't automatically delete repository files if they are present

### PR DESCRIPTION
**Pull request by zeripath**
Prior to this PR Gitea would delete any repository files if they are
present during creation or migration. This can in certain circumstances
lead to data-loss and is slightly unpleasant.

This PR provides a mechanism for Gitea to adopt repositories on creation
and otherwise requires an explicit flag for deletion.

PushCreate is slightly different - the create will cause adoption if
that is allowed otherwise it will delete the data if that is allowed.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/repository/create.go [hunk 1]
```diff
@@ -39,26 +40,70 @@ func CreateRepository(doer, u *models.User, opts models.CreateRepoOptions) (_ *m
...
 	}
 
-	err = models.WithTx(func(ctx models.DBContext) error {
-		if err = models.CreateRepository(ctx, doer, u, repo); err != nil {
+	overwriteOrAdopt := (!opts.IsMirror && opts.AdoptPreExisting && setting.Repository.AllowAdoptionOfUnadoptedRepositories) ||
+		(opts.OverwritePreExisting && setting.Repository.AllowOverwriteOfUnadoptedRepositories)
+
+	repoPath := models.RepoPath(u.Name, repo.Name)
+	if !overwriteOrAdopt && com.IsExist(repoPath) {
+		log.Error("Files already exist in %s and we are not going to adopt or delete.", repoPath)
```
Comment (a1012112796) [inline modules/repository/create.go]: I don't think writing an error log here is necessary.

File: modules/repository/create.go [hunk 2]
```diff
@@ -13,10 +13,11 @@ import (
 	"code.gitea.io/gitea/modules/git"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
+	"github.com/unknwon/com"
```
Comment (lunny) [inline modules/repository/create.go:17]: blank line.

**Comment by lunny** [comment]
How about not give the option to overwrite the exist repository and just returned error?

File: modules/repository/create.go (hunk 2 above)
Comment (zeripath) [reply modules/repository/create.go:17]: done

**Review by 6543 — APPROVED** [review]

File: models/repo.go [hunk 3]
```diff
@@ -1023,24 +1023,30 @@ func CheckCreateRepository(doer, u *User, name string) error {
 	} else if has {
```
Comment (lunny) [inline models/repo.go:1063]: I think we have checked if it exists here.

Comment (zeripath) [reply models/repo.go:1063]: Not in the same way.
isRepositoryExist requires that there is a dB entry and the repo files exists.

The next test just checks if there are files present.

**Comment by lafriks** [comment]
Please resolve conflicts

**Comment by 6543** [comment]
swagger lint

**Comment by zeripath** [comment]
OK Swagger just needs to be run as part of our generate. Why aren't we doing that?

**Comment by zeripath** [comment]
I'm serious - I don't understand why the swagger generate can't be done on generate. We can ensure that we generate a valid swagger description.

**Review by mrsdizzie — APPROVED** [review]

File: modules/migrations/gitea.go [hunk 4]
```diff
@@ -108,13 +108,14 @@ func (g *GiteaLocalUploader) CreateRepo(repo *base.Repository, opts base.Migrate
...
-			IsMirror:       opts.Mirror,
-			Status:         models.RepositoryBeingMigrated,
+			Name:                 g.repoName,
+			Description:          repo.Description,
+			OriginalURL:          repo.OriginalURL,
+			GitServiceType:       opts.GitServiceType,
+			IsPrivate:            opts.Private,
+			IsMirror:             opts.Mirror,
+			Status:               models.RepositoryBeingMigrated,
+			OverwritePreExisting: opts.OverwritePreExisting && (g.doer.IsAdmin || setting.Repository.AllowOverwriteOfUnadoptedRepositories),
```
Comment (lunny) [inline modules/migrations/gitea.go:118]: Why an admin could not follow the setting?

File: modules/repository/init.go [hunk 5]
```diff
@@ -176,9 +180,45 @@ func checkInitRepository(repoPath string) (err error) {
...
+	repo.IsEmpty = false
+
+	repo.DefaultBranch = setting.Repository.DefaultBranch
+	if len(opts.DefaultBranch) > 0 {
+		repo.DefaultBranch = opts.DefaultBranch
+		gitRepo, err := git.OpenRepository(repo.RepoPath())
+		if err != nil {
+			return fmt.Errorf("openRepository: %v", err)
+		}
+		if err = gitRepo.SetDefaultBranch(repo.DefaultBranch); err != nil {
```
Comment (lunny) [inline modules/repository/init.go:219]: defer gitRepo.Close()

**Comment by lunny** [comment]
I don't like this PR now because I think it makes the process more complicated. Maybe we should add a new feature which is only a button `adopt projects`, then it will scan all the repositories root folder and automatically check if it exists on the database, if not then create it. If it's scanning, the button could not be clicked.

For this PR we can just return the error if the repository folder exists.

File: modules/migrations/gitea.go (hunk 4 above)
Comment (zeripath) [reply modules/migrations/gitea.go:118]: ? This allows Site Admins to explicitly overwrite pre-existing repositories.

File: modules/repository/init.go (hunk 5 above)
Comment (zeripath) [reply modules/repository/init.go:219]: done

**Comment by zeripath** [comment]
Any UI that involves adopting repositories is going to have to provide the information like:

* Visibility
* Description
* Issue labels
* Default branch

it's then going to require a UI to search to find repositories to adopt / overwrite.

**Comment by lunny** [comment]
You can change them after adopting them. Then you have resued the previous UI and features. And also this will resolve batching `adopt projects` feature.

**Review by 6543 — CHANGES_REQUESTED** [review]
lunny's suggestion sounds resonable

**Comment by jolheiser** [comment]
Under the context that this is primarily intended to stop unnecessary data loss (adoption/overwriting being a secondary effect of that), I think this PR is great.

I think the code it introduces could be used in a later PR to allow for batch adoption of repositories on disk.

**Review by 6543 — APPROVED** [review]
this is nice now :)

**Pull request closed without merging**